Allow sending signals to processes inside a pot

grembo opened the pull request:
`pot signal` allows sending signals to processes running inside a pot.

Non-persistent pots will receive the signal by default on their
main process (the one started by `pot start`), non-persistent pots can
still receive signals as specified by either `-P pid` or `-m pattern`.

Example of use:

    pot signal -s KILL -p mypot
    pot signal -p mypot -f -m "sleep"

grembo commented:
@pizzamig Someone :) needs to write unit tests.

pizzamig commented on share/pot/signal.sh:
```diff
@@ -0,0 +1,124 @@
+		_tmpfile=$(mktemp \
+		  "${POT_TMP:-/tmp}/pot_sigpid_${_pname}${POT_MKTEMP_SUFFIX}"
+		  ) || exit 1
+		echo "$_pid" >"$_tmpfile" || exit 1
```
Consider using `return` instead of `exit`
Alternatively, you can use `${EXIT}`, that assume the value of `return` in unit tests and `exit` otherwise

grembo replied:
I got this from some other place where we explicitly exit in case of a low level failure. Maybe a bad idea?

pizzamig commented on share/pot/signal.sh:
```diff
@@ -0,0 +1,124 @@
+	else
+		_info "Sending $_signal to main process of pot $_pname"
+		pkill "-$_signal" -j "$_pname" \
+		  -F "${POT_TMP:-/tmp}/pot_main_pid_${_pname}"
```
This file exist only for non-persistent pot.
Should we add some checks like if pot is non-persitent or that the file exists?

grembo replied:
Let's check for the persistent setting. For the exceptional case that the file was deleted and/or someone bothered to change the persistent configuration while the pot is running, pkill will give a very human readable error message.

pizzamig commented on share/pot/signal.sh:
```diff
@@ -0,0 +1,124 @@
+signal-help()
+{
+	cat <<-"EOH"
+	pot signal [-hv] [-s signal] [-P pid] [-m pattern]-p pot
```
missing space before `-p`

pizzamig commented on share/pot/signal.sh:
```diff
@@ -0,0 +1,124 @@
+	  -P pid : the pid inside the pot to send the signal to.
+	           for non-persistent pots this defaults to
+	           the pid of the main process
+          -m pattern : A pattern to match (overrides -P)
```
It seems that there is some misalignment for `-m pattern`.
Tabs vs Space, maybe?

pizzamig commented on share/pot/start.sh:
```diff
@@ -591,12 +592,17 @@ _js_start()
 	sleep 0.5
 	pkill -f -j "$_pname" "^sleep 1234$"
 
+        if [ "$_persist" = "NO" ]; then
```
Another Tab vs Space?

pizzamig reviewed (changes requested):
Few remarks I've found

pizzamig commented:
I can write tests for it, but I will also reorganize the code in functions, so I can test those functions instead of the entire command, it makes mocking way easier

grembo commented:
@pizzamig I think I addressed all feedback. I added a few more features/improvements.
> I can write tests for it, but I will also reorganize the code in functions, so I can test those functions instead of the entire command, it makes mocking way easier

I restructured the code, so the new version gives you the following functions you can write unit tests for:

- `_get_signal_names()`
- `_validate_signal_name()`
- `_validate_pid()`
- `_send_signal()`

(you should be able to stub "pgrep" and "pkill", therefore I didn't write callbacks/overrides for them)

pizzamig reviewed (approved):
It seems good to me!

pizzamig commented:
You can merge it, I'll add unit tests once it's merged :)

grembo commented:
Thanks for reviewing
